packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx

```diff
@@ -68,98 +91,138 @@ export const quantityTypesSelectionOptions: SelectOption<QuantityType>[] = [
+  // ECProperty source schema:class
+  schema: string;
+  className: string;
+  categoryLabel: string;
+  // Property Type
+  type: DataType;
+  // The actual ECClass name of the property
+  actualECClassName: string;
+  // The parent class of the property
+  parentPropertyClassName: string | undefined;
+  // ECProperty type traversal
+  propertyTraversal: Array<string>;
+  // The type of primitive navigation from Presentation
+  primitiveNavigationClass: string;
+  key: string;
 }
 
-const extractPrimitive = (
-  propertiesField: PropertiesField,
-  classToPropertiesMapping: Map<string, Property[]>,
-  navigation?: NavigationProperty
-) => {
-  // There are rare cases which only happens in multiple selections where it returns more than one.
-  // This also checks if this property comes from a navigation property
-  const className =
-    navigation?.rootClassName ??
-    propertiesField.properties[0].property.classInfo.name;
-  // Sometimes class names are not defined. Type error. Not guaranteed.
-  if (!className) {
-    return;
-  }
-
-  if (!classToPropertiesMapping.has(className)) {
-    classToPropertiesMapping.set(className, []);
+const convertType = (type: string): DataType => {
+  switch (type) {
+    case "int":
+    case "enum":
+    case "long":
+      return DataType.Integer;
+    case "boolean":
+      return DataType.Boolean;
+    case "double":
+      return DataType.Double;
+    case "string":
+      return DataType.String;
+    case "number":
+      return DataType.Number;
+    default:
+      return DataType.Undefined;
   }
+};
 
-  // Gets property name. Appends path if from navigation.
-  const propertyName = navigation
-    ? `${navigation.navigationName}.${propertiesField.properties[0].property.name}`
-    : propertiesField.properties[0].property.name;
-
-  const label = navigation
-    ? `${propertiesField.label} (${navigation?.navigationName})`
-    : propertiesField.label;
-
-  // Ignore hardcoded BisCore navigation properties
-  if (propertiesField.type.typeName === "navigation") {
-    return;
-  } else {
-    classToPropertiesMapping.get(className)?.push({
-      name: propertyName,
-      label,
-      type: propertiesField.properties[0].property.type,
-    });
-  }
+const extractPrimitive = (
+  propertyTraversal: Array<string>,
+  propertyField: PropertiesField
+): PropertyMetaData => {
+  const propertyName = propertyField.properties[0].property.name;
+  const label = propertyField.label;
+  //  It belongs to this parent class
+  const parentPropertyClassName = propertyField.parent?.contentClassInfo.name;
+  const primitiveNavigationClass = propertyField.properties[0].property.navigationPropertyInfo?.classInfo.name ?? "";
+  // Presentation treats primitive navigation classes as longs. Handling this special case.
+  const type = primitiveNavigationClass ? DataType.String : convertType(propertyField.properties[0].property.type);
+  const actualECClassName = propertyField.properties[0].property.classInfo.name;
```

Comment:
In what cases will the `propertyField.properties` array contain more than 1 entry?

Reply:
I believe it happens during property merging. Forgot to consider this case. https://github.com/iTwin/itwinjs-core/blob/master/docs/presentation/content/Terminology.md#property-merging
I wrote the logic to deal with this but I cannot find an iModel where this edge case occurs.